apis/storage/v1alpha3/types.go

```diff
@@ -597,6 +597,13 @@ type BucketUpdatableAttrs struct {
 	// for valid values.
 	PredefinedDefaultObjectACL string `json:"predefinedDefaultObjectAcl,omitempty"`
 
+	// PublicAccessPrevention is the setting for the bucket's
+	// PublicAccessPrevention policy, which can be used to prevent public access
+	// of data in the bucket. See
+	// https://cloud.google.com/storage/docs/public-access-prevention for more
+	// information.
+	PublicAccessPrevention string `json:"publicAccessPrevention,omitempty"`
```

Comment:
> I figured exposing the raw int enum from the google storage API was undesirable, but I can't see e.g a webhook where the allowed values can be verified early.

Based on your code, I believe we should have three options for this field: `inherited`, `enforced` or `unknown` (do we need the `unknown` value?). Please correct me if I'm wrong.

The unknown value is the default value for the API if you don't specify the PublicAccessPrevention, right?

You can use [kubebuilder marker](https://book.kubebuilder.io/reference/markers/crd-validation.html) to validate the possible values. Check [this example](https://github.com/crossplane-contrib/provider-gcp/blob/30ff5893b2a117ce20135a0edb5643ca896681a0/apis/storage/v1alpha3/types.go#L722) about how to define validation:Enum markers

---

Besides that, I believe that this field is optional, so we should add the `// +optional` marker and change the field type to be a pointer.

```suggestion
	// PublicAccessPrevention is the setting for the bucket's
	// PublicAccessPrevention policy, which can be used to prevent public access
	// of data in the bucket. See
	// https://cloud.google.com/storage/docs/public-access-prevention for more
	// information.
	//
	// +optional
	PublicAccessPrevention *string `json:"publicAccessPrevention,omitempty"`
```

Reply:
Since another PR (https://github.com/crossplane-contrib/provider-gcp/pull/470) has bumped the dependencies version I'm rebasing this PR based the current master branch. Hope this is ok.

Regarding the optionality of the field, I believe having it as the empty string by default is the correct option. Looking at the bucket attributes in the library it is an integer (https://github.com/googleapis/google-cloud-go/blob/main/storage/bucket.go#L345, defaulting to 0, which is the same as "unknown": https://github.com/googleapis/google-cloud-go/blob/main/storage/bucket.go#L485)

> The unknown value is the default value for the API if you don't specify the PublicAccessPrevention, right?

It's the default value both sent to and received from the API when the PAP is not specified (which is the case prior to this PR).

The kubebuilder markers is a good idea :) I've added some kubebuilder markers to reflect that it's an enumeration with the possible values for the field. I still specify it as optional as there (possibly) is a sensible default value for this field. Does this make sense to you?

Reply:
Did some testing using this PR:

- Creating a bucket with no `publicAccessPrevention` in the `.spec`: Results in the value `inherited` on my GCP account and the value "inherited" being reflected back to the `.spec.publicAccessPrevention` field.
- Creating a bucket with `publicAccessPrevention: ""` in the `.spec`: Results in the value `inherited` on my GCP and the value "inherited" being reflected back to the `.spec.publicAccessPrevention` field.account.

I'm adding a commit to this PR that makes this field optional as you suggested, but for it to make sense I also needed to change the Observe function a little to allow for the setting to only exist in the GCP API and not in the CR spec (see https://github.com/crossplane-contrib/provider-gcp/pull/467/commits/c3f0076a330ace04ac16888bcbc2a5b77f8093e1#diff-b25ecc0642483145f66b0c3e2fc414c15b740c9e4f9aca54f8bc2aa193488409R145-R154). Please let me know what you think :)